packages/extension/src/ui/action/views/network-assets/components/custom-evm-token.vue

```diff
@@ -0,0 +1,481 @@
+const isFocus = ref(false);
+
+const isValidAddress = computed(() => {
+  if (contractAddress.value) {
+    return isAddress(contractAddress.value);
+  }
+
+  return false;
+});
+
+watch([contractAddress, props], async () => {
+  notTokenAddress.value = false;
+  tokenInfo.value = undefined;
+  accountBalance.value = undefined;
+
+  if (!props.network.customTokens) {
+    close();
+  }
+
+  if (isValidAddress.value) {
+    const api = (await props.network.api()) as API;
+
+    const info = await api.getTokenInfo(contractAddress.value!);
+
+    if (info.name !== "Unknown") {
+      let icon = props.network.icon;
+
+      try {
+        const controller = new AbortController();
+
+        const timeoutId = setTimeout(() => controller.abort(), 15_000);
+
+        const res = await fetch(
+          `https://api.coingecko.com/api/v3/coins/${props.network.coingeckoID}/contract/${contractAddress.value}`,
+          { signal: controller.signal }
+        );
+
+        clearTimeout(timeoutId);
+
+        if (res.ok) {
+          const data = await res.json();
+
+          const image = data?.image?.small;
+
+          if (image) {
+            icon = `https://img.mewapi.io/?image=${image}`;
+          }
+        }
+      } catch (error) {
+        // Icon already defaults to network icon
+      }
+
+      if (props.address !== "") {
+        const erc20Token = new Erc20Token({
+          name: info.name,
+          symbol: info.symbol,
+          decimals: info.decimals,
+          icon,
+          contract: contractAddress.value!,
+        });
+
+        try {
+          const balance = await erc20Token.getLatestUserBalance(
+            api,
+            props.address
+          );
+          accountBalance.value = balance;
+        } catch {
+          // Don't set balance
+        }
+      }
+
+      tokenInfo.value = {
+        address: contractAddress.value as `0x${string}`,
+        icon,
+        type: TokenType.ERC20,
+        ...info,
+      };
+    } else {
+      notTokenAddress.value = true;
+    }
+  }
+});
+
+const changeFocus = (val: FocusEvent) => {
+  isFocus.value = val.type === "focus";
+};
+
+const close = () => {
+  emits("update:close");
+};
+
+const addToken = async () => {
+  if (isValidAddress.value && tokenInfo.value) {
+    const inserted = await tokensState.addErc20Token(
+      props.network.name,
+      toRaw(tokenInfo.value)
+    );
+
+    if (inserted) {
+      const newAsset: AssetsType = {
+        name: tokenInfo.value.name,
+        symbol: tokenInfo.value.symbol,
+        balance: accountBalance.value ?? "0",
+        balancef: formatFloatingPointValue(
+          fromBase(accountBalance.value ?? "0", tokenInfo.value.decimals)
+        ).value,
+        contract: tokenInfo.value.address,
+        balanceUSD: 0,
+        balanceUSDf: "0",
+        value: "0",
+        valuef: "0",
+        decimals: tokenInfo.value.decimals,
+        sparkline: "",
+        priceChangePercentage: 0,
```

Comment:
these should be updated if there is a coingeckotoken

Reply:
@kvhnuke  Should coingeckoID be added to `CustomToken` if it exists?

Reply:
Yea that will be better